EAP7-1596: new tests for MP JWT 1.2

tommaso-borgato opened the pull request:
MP JWT - shall be dealt in https://issues.redhat.com/browse/EAP7-1596

CC @jstourac, @honza-kasik 

`microprofile-jwt` runs successfully on my laptop:

```
...
[INFO] 
[INFO] Results:
[INFO] 
[WARNING] Tests run: 64, Failures: 0, Errors: 0, Skipped: 9
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:06 min
[INFO] Finished at: 2021-02-10T09:44:46+01:00
[INFO] ------------------------------------------------------------------------
```

honza-kasik commented on microprofile-jwt/src/main/resources/basic-authentication.web.xml:
```diff
@@ -0,0 +1,6 @@
+    <login-config>
+        <auth-method>BASIC</auth-method>
+    </login-config>
+</web-app>
```
Please add newline.

tommaso-borgato replied:
done

honza-kasik commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/activation/AnnotationActivationCompatibilityTest.java:
```diff
@@ -70,6 +71,8 @@ public static void beforeClass() throws URISyntaxException {
     public void testSameTokenReceived(@ArquillianResource URL url) {
         JsonWebToken token = new JwtHelper(keyTool).generateProperSignedJwt();
 
+        System.out.println("\n\nJWT:\t" + token.getRawValue() + "\n\n");
```
Possibly a forgotten debug line...

honza-kasik commented on microprofile-jwt/src/main/java/org/jboss/eap/qe/microprofile/jwt/testapp/servlet/SecuredServlet.java:
```diff
@@ -0,0 +1,50 @@
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ *
```
Please document purpose of this servlet.

honza-kasik commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JsonWebTokenApiTest.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    /**
+     * A valid signed JWT token is sent to the server and, in the JAX-RS endpoint, the new API method `getClaim` is
```
Please use proper test documentation https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/master/CONTRIBUTING.md#test-documentation

honza-kasik commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JsonWebTokenApiTest.java:
```diff
@@ -0,0 +1,87 @@
+     * A valid signed JWT token is sent to the server and, in the JAX-RS endpoint, the new API method `getClaim` is
+     * used to extract a claim
+     * 
+     * @param url
```
If documentation won't be provided, please don't include the param or exception

honza-kasik commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/deployment/SessionScopedClaimInjectionTest.java:
```diff
@@ -0,0 +1,47 @@
+import org.junit.runner.RunWith;
+
+/**
+ * MP-JWT implementations are required to generate a `javax.enterprise.inject.spi.DeploymentException` for a claim value
```
Please use `{@code foobar}` instead of backtics.

tommaso-borgato replied:
@honza-kasik done

honza-kasik commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/MissingJwtTest.java:
```diff
@@ -0,0 +1,101 @@
+                .statusCode(200);
+    }
+
+    @Test
```
Please document the test cases https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/master/CONTRIBUTING.md#test-documentation

honza-kasik replied:
Still missing

honza-kasik commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/MissingJwtTest.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * @tpTestDetails Supply an expired ({@code exp} set in past) JWT to server
+     * @tpPassCrit Expired JWT is rejected and user receives 401/Unauthorized.
+     * @tpSince EAP 7.4.0.CD19
```
This doc doesn't describe this test.

honza-kasik replied:
` * @tpSince EAP 7.4.0.CD19` - this test is here from CD23 AFAIK

honza-kasik commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/security/rbac/cookie/JwsCookieTokenTestCase.java:
```diff
@@ -0,0 +1,282 @@
+     * @param url
+     * @throws Exception
+     */
+    @Test
```
Please document tests properly https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/master/CONTRIBUTING.md#test-documentation

honza-kasik commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/security/rbac/jwe/JweTestCase.java:
```diff
@@ -0,0 +1,307 @@
+                        "key.private.pkcs8.pem");
+    }
+
+    @Test
```
Please include proper test documentation https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/master/CONTRIBUTING.md#test-documentation

honza-kasik commented on tooling-mp-jwt-auth-tool/pom.xml:
```diff
@@ -15,8 +15,19 @@
         </dependency>
+        <dependency>
+            <groupId>org.bitbucket.b_c</groupId>
+            <artifactId>jose4j</artifactId>
```
`<scope>test</scope>`?

tommaso-borgato replied:
used `<scope>provided</scope>` because `jose4j` is used in e.g. JweHelper

honza-kasik commented on tooling-mp-jwt-auth-tool/src/test/resources/pki/gen-keys.sh:
```diff
@@ -0,0 +1,59 @@
+function gen_key_pair() {
```
Why there are generated keypairs in tooling module? If those are test resources, please move them to module containing MP tests.

tommaso-borgato replied:
those keys are used in tests e.g. `RsaKeyHelperTest`

honza-kasik replied:
So these keys and their generation are now duplicated too?

tommaso-borgato replied:
@honza-kasik the script is the same but the generated keys are different and are used in different tests & modules

honza-kasik replied:
@tommaso-borgato This means that we have to maintain same code on two places. This needs to be solved.

tommaso-borgato replied:
@honza-kasik fixed

honza-kasik commented on tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/JwsHelper.java:
```diff
@@ -0,0 +1,191 @@
+     * @return
+     * @throws Exception
+     */
+    public JsonWebSignature generateProperSignedJwt(
```
What is the reason for this method when `org.jboss.eap.qe.microprofile.jwt.auth.tool.JwtHelper#generateProperSignedJwtWithClaims` exist?

tommaso-borgato replied:
I can't say exactly because it was there before my modifications .... I'm responsible for JwsHelper#generateProperSignedJwt which basically emulates JwtHelper#generateProperSignedJwtWithClaims but using `jose4j`

honza-kasik replied:
@tommaso-borgato I understand it uses `jose4j` but why it is used here now?

tommaso-borgato replied:
@honza-kasik it's the counterpart of the former method `generateProperSignedJwtWithClaims` but using `jose4j`

honza-kasik replied:
@tommaso-borgato Sorry, I might worded that wrong. What I meant is that it is obvious jose4j is used. But if we are going to introduce it, it should be documented why.

tommaso-borgato replied:
@honza-kasik I deprecated `generateProperSignedJwtWithClaims` and added some documentation to justify it

honza-kasik replied:
@tommaso-borgato IMHO we are still missing docs why jose4j was added (the reason for deprecation)

tommaso-borgato replied:
@honza-kasik just point me to the correct location for adding that .... and I'll add the docs

tommaso-borgato replied:
basically it' here https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/release-notes.asciidoc#release-notes-for-microprofile-jwt-12:
```
TCK tests now use Jose4J to sign and encrypt the tokens.
```

honza-kasik replied:
@tommaso-borgato From point of view of someone who just opened testsuite it looks like this:

* There's this and `org.jboss.eap.qe.microprofile.jwt.auth.tool.JwtHelper#generateProperSignedJwtWithClaims`
* `org.jboss.eap.qe.microprofile.jwt.auth.tool.JwtHelper#generateProperSignedJwtWithClaims` is marked as deprecated and jose4j is mentioned
* Explanation how is jose4j reason for deprecation os nowhere to be found.

> TCK tests now use Jose4J to sign and encrypt the tokens.

We migrated some TCK tests here?

tommaso-borgato replied:
`Jose4J` makes working with JWT stuff easier and alleviates the burden of rewriting and maintaining those functionalities .... just like any other library

tommaso-borgato replied:
@honza-kasik is the javadoc https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/08968f5d3bab6a3f30d2a77606921a883309bd41/tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/JwtHelper.java#L137 enough or shall any more refactoring is needed?

tommaso-borgato replied:
@honza-kasik I added the info you added here https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/32a6fd6b335be4aeec2d8d0d38c7aba8c429297a/tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/JwtHelper.java#L138; I added the things actually used used, let me know if it's enough or if I need to add some more extensive list

honza-kasik replied:
Thank you @tommaso-borgato regarding documentation this is sufficient now.

However the introduced changes still bring a lot of redundancy. Since the re-write would mean a big investment and possible planning issues it will be merged with the redundant code present. Optimal solution would mean extending existing tooling with missing features such as JWT encryption and elliptic curve support. If this will present increased maintenance cost in the future. It can be "fixed" later.

honza-kasik commented on tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/AbstractKeyLoader.java:
```diff
@@ -0,0 +1,73 @@
+ * @param <PUBLIC_KEY> : Public Key interface for the algorithm in use
+ * @author tborgato
+ */
+public abstract class AbstractKeyLoader<PRIVATE_KEY, PUBLIC_KEY> {
```
Isn't this redundant to `org.jboss.eap.qe.microprofile.jwt.auth.tool.RsaKeyTool`?

tommaso-borgato replied:
probably yes, but I needed it in order to handle the new EC algorithm

tommaso-borgato replied:
I mean I had to add code for the new EC algorithm somewhere and, since we'll probably progressively introduce `jose4j`, I preferred to keep it separated in order ease pruning in some future .... I can try and merge the two but it would take some time since I'd have to revise and test all places where it's used .... I can do it but if you agree I'll do it in some subsequent MR since this RFE needs to be ready by end of next week

honza-kasik replied:
From my point of view is we now have bunch of redundant code which does exactly the same thing. I'm really not fan of subsequent MRs because then it will be another record in issues and when merged, this redundancy will bring the code maintainability down.

tommaso-borgato replied:
@honza-kasik doing the complete migration in this MR would take a lot of time .... I can deprecated things in `RsaKeyTool` to make the intent evident ...

honza-kasik replied:
@tommaso-borgato Yes please, deprecate it then and properly document it.

tommaso-borgato replied:
@honza-kasik done

honza-kasik replied:
Thanks! LGTM

honza-kasik commented:
Thanks @tommaso-borgato ! There are three issues:

* Proper test documentation is missing, see https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/master/CONTRIBUTING.md#test-documentation please
* Added tooling seems redundant. What was the reason for not extending existing tooling and creating new instead?
* There is formatting issue, see Travis output.

jstourac commented on microprofile-jwt/pom.xml:
```diff
@@ -15,6 +15,8 @@
             <groupId>org.eclipse.microprofile</groupId>
             <artifactId>microprofile</artifactId>
+            <!-- TODO: remove once all microprofile specs will be updated to MP4 -->
+            <version>4.0.1</version>
```
Just a note - if you rebase against current MP4 branch, you should be able to get rid of this already since there is performed update for all modules.

tommaso-borgato commented:
@honza-kasik thank you for your review .... I guess I fixed everything .... please note I added `.withEnvVar("DB_VENDOR", "h2")` to the KEYCLOAK docker container so it uses an embedded database

tommaso-borgato commented:
@honza-kasik https://issues.redhat.com/browse/EAP7-1596?focusedCommentId=15812034&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15812034

honza-kasik commented:
Hi @tommaso-borgato. The test documentation per https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/master/CONTRIBUTING.md#test-documentation is still missing on lot of places. I also responded to code redundancy issues. Thank you and sorry for late reply.

tommaso-borgato commented:
@honza-kasik I just added the missing javadocs ..... now looking into the rest ....

honza-kasik commented on microprofile-jwt/src/test/resources/pki/gen-keys.sh:
```diff
@@ -1,15 +1,63 @@
+    openssl rsa -in "${PREFIX}.private.pkcs8.pem" -pubout -outform PEM -out "${PREFIX}.public.pem"
+
+    popd
+  elif [[ "${ALGORITHM}" == "ES256" ]]; then
```
Maybe just a styling issue, but maybe this can be split to three functions instead since the amount of shared code is low.

tommaso-borgato replied:
@honza-kasik good point .... leaving a comment

honza-kasik commented on tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/AbstractKeyLoader.java:
```diff
@@ -0,0 +1,73 @@
+
+    protected String algorithm;
+
+    public PRIVATE_KEY readPrivateKey(final URI pkcs8Key) throws Exception {
```
Please add JavaDoc

tommaso-borgato replied:
@honza-kasik added

honza-kasik commented on tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/AbstractKeyLoader.java:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    public PUBLIC_KEY readPublicKey(final URI pemKey) throws Exception {
```
Please add JavaDoc

tommaso-borgato replied:
@honza-kasik added

honza-kasik commented on tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/JweHelper.java:
```diff
@@ -0,0 +1,289 @@
+        return ecHelper;
+    }
+
+    public JsonWebEncryption generateJwe(
```
Method with more than 4 arguments is seen as bad practice. IMHO a builder pattern would be very useful here.

tommaso-borgato replied:
@honza-kasik good point, do you want me to change it in this PR?

tommaso-borgato replied:
@honza-kasik added TODO for next PR

honza-kasik commented on tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/JweHelper.java:
```diff
@@ -0,0 +1,289 @@
+        return String.join(".", jweTokens);
+    }
+
+    public JwtContext extractNestedJwt(
```
Please add a JavaDoc

tommaso-borgato replied:
@honza-kasik added javadoc on all methods

honza-kasik commented on tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/JwsHelper.java:
```diff
@@ -0,0 +1,192 @@
+     * @return {@link JsonWebSignature} representing the JWS
+     * @throws Exception in case anything goes wrong while generating the JWS
+     */
+    public JsonWebSignature generateProperSignedJwt(
```
A builder pattern would better option here.

tommaso-borgato replied:
@honza-kasik good idea, do you want me to modify it in this PR?

honza-kasik replied:
@tommaso-borgato This can be done in next one, to keep this "short".

honza-kasik commented on tooling-mp-jwt-auth-tool/src/test/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/EcKeyHelperTest.java:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * @tpTestDetails Reads a private key from a pem file
+     * @tpPassCrit Expect no Exceptions
+     * @tpSince EAP 7.4.0.CD19
```
This was added later.

tommaso-borgato replied:
@honza-kasik yes I added them last Friday

honza-kasik replied:
@tommaso-borgato 😄 I meant in CD23...

tommaso-borgato replied:
@honza-kasik :rofl: you r right .... fixed in all tests in the same package

honza-kasik commented on tooling-mp-jwt-auth-tool/src/test/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/EcKeyHelperTest.java:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * @tpTestDetails Reads a public key from a pem file
+     * @tpPassCrit Expect no Exceptions
+     * @tpSince EAP 7.4.0.CD19
```
This was added later.

tommaso-borgato replied:
@honza-kasik yes I added them last Friday

tommaso-borgato replied:
@honza-kasik CD23 fixed

honza-kasik reviewed (changes requested):
Thank you @tommaso-borgato ! Main issues regarding code redundancy were resolved, though there are other issues. Please see my comments.

tommaso-borgato commented on microprofile-jwt/src/main/java/org/jboss/eap/qe/microprofile/jwt/testapp/jaxrs/SessionScopedJaxRsEndpoint.java:
```diff
@@ -0,0 +1,36 @@
+import org.jboss.eap.qe.microprofile.jwt.testapp.Endpoints;
+
+/**
+ * MP-JWT implementations are required to generate a `javax.enterprise.inject.spi.DeploymentException` for a claim value
```
@honza-kasik fixed

tommaso-borgato commented on microprofile-jwt/src/main/java/org/jboss/eap/qe/microprofile/jwt/testapp/servlet/SecuredServlet.java:
```diff
@@ -0,0 +1,51 @@
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Used to compare behaviour between `MP-JWT` and `BASIC` authentication
```
@honza-kasik apparently I use back-tics a lot since I learned the keyboard shortcut .... I fixed also other javadocs where I used it

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JsonWebTokenApiTest.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    /**
+     * A valid signed JWT token is sent to the server and, in the JAX-RS endpoint, the new API method `getClaim` is
```
@honza-kasik fixed

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/MissingJwtTest.java:
```diff
@@ -0,0 +1,116 @@
+
+/**
+ * The @see org.eclipse.microprofile.auth.LoginConfig annotation provides the same information as the web.xml login-config
+ * element; this tests verifies that, when present, the following behaviours are the same for both `MP-JWT` and `BASIC`
```
@honza-kasik fixed

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/MissingJwtTest.java:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * @tpTestDetails Supply no JWT to server
+     * @tpPassCrit Expect HTTP/200
+     * @tpSince EAP 7.4.0.CD19
```
@honza-kasik fixed

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/MissingJwtTest.java:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * @tpTestDetails Supply invalid BASIC authentication to server
+     * @tpPassCrit Expect 401/Unauthorized
+     * @tpSince EAP 7.4.0.CD19
```
@honza-kasik  fixed

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/MissingJwtTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    /**
+     * @tpTestDetails Supply no BASIC authentication to server
```
@honza-kasik fixed

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/MissingJwtTest.java:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * @tpTestDetails Supply no BASIC authentication to server
+     * @tpPassCrit Expect HTTP/200
+     * @tpSince EAP 7.4.0.CD19
```
@honza-kasik fixed

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/deployment/SessionScopedClaimInjectionTest.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Deployment(name = DEPLOYMENT_NAME, managed = false)
+    public static WebArchive createDeployment() {
+        return ShrinkWrap.create(WebArchive.class, SessionScopedClaimInjectionTest.class.getSimpleName() + ".war")
```
@honza-kasik tnks .... copy and paste leftover ....

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/deployment/SessionScopedClaimInjectionTest.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     * @tpTestDetails Deploys and app containing a @SessionScoped bean with a claim value injection
+     * @tpPassCrit Expect deployment to fail wit javax.enterprise.inject.spi.DeploymentException
```
@honza-kasik fixed

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/security/rbac/cookie/JwsCookieTokenTestCase.java:
```diff
@@ -0,0 +1,284 @@
+package org.jboss.eap.qe.microprofile.jwt.security.rbac.cookie;
```
@honza-kasik moved

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/security/rbac/cookie/JwsCookieTokenTestCase.java:
```diff
@@ -0,0 +1,284 @@
+@RunAsClient
+@RunWith(Arquillian.class)
+@ServerSetup(EnableJwtSubsystemSetupTask.class)
+public class JwsCookieTokenTestCase {
```
@honza-kasik good point .... adding a TODO

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/security/rbac/cookie/JwsCookieTokenTestCase.java:
```diff
@@ -0,0 +1,284 @@
+
+    /**
+     * @tpTestDetails Test invokes a JAX-RS endpoint which expects authentication and authorization provided by an
+     *                RS encoded JWS token supplied as cookie containing the required claims for authorization;
```
@honza-kasik you r right: it's "signed" not encoded

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/MissingJwtTest.java:
```diff
@@ -0,0 +1,116 @@
+
+    /**
+     * @tpTestDetails Supply no BASIC authentication to server
+     * @tpPassCrit Expect HTTP/200
```
@honza-kasik modified

tommaso-borgato commented:
@honza-kasik I guess I've addressed all the issues you pointed out;

JWT tests run successfully:
```
[INFO] 
[INFO] Results:
[INFO] 
[WARNING] Tests run: 64, Failures: 0, Errors: 0, Skipped: 9
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:04 min
[INFO] Finished at: 2021-02-22T15:46:40+01:00
[INFO] ------------------------------------------------------------------------

```

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/security/rbac/jwe/JweTestCase.java:
```diff
@@ -0,0 +1,335 @@
+import org.junit.runner.RunWith;
+
+/**
+ * MP JWT 1.2 introduced the support for Support for decrypting JWT tokens which have been encrypted using RSA-OAEP and
```
@honza-kasik fixed

tommaso-borgato commented on microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/security/rbac/jwe/JweTestCase.java:
```diff
@@ -0,0 +1,335 @@
+ * this test verifies that:
+ *
+ * <ul>
+ * <li>JWE token are correctly used for granting/denying authentication and authorization</li>
```
@honza-kasik fixed

honza-kasik commented:
Thanks @tommaso-borgato . Apart bash script code redundancy (see my latest comment) I didn't find any serious issue.

tommaso-borgato commented:
@honza-kasik fixed the redundancy issue
